algorithms/linfa-clustering/src/optics/algorithm.rs

```diff
@@ -0,0 +1,291 @@
+use crate::optics::hyperparameters::{OpticsHyperParams, OpticsHyperParamsBuilder};
+use float_ord::FloatOrd;
+use hnsw::{Params, Searcher, HNSW};
+use linfa::traits::Transformer;
+use linfa::Float;
+use ndarray::{ArrayBase, ArrayView, ArrayView1, Axis, Data, Ix1, Ix2};
+use ndarray_stats::DeviationExt;
+#[cfg(feature = "serde")]
+use serde_crate::{Deserialize, Serialize};
+use space::MetricPoint;
+use std::cmp::{max, Ordering};
+use std::collections::BTreeSet;
+
+/// Implementation of euclidean distance for ndarray
+struct Euclidean<'a, F>(ArrayView1<'a, F>);
+
+impl<F: Float> MetricPoint for Euclidean<'_, F> {
+    fn distance(&self, rhs: &Self) -> u32 {
+        let val = self.0.l2_dist(&rhs.0).unwrap();
+        space::f64_metric(val)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_crate")
+)]
+/// OPTICS (Ordering Points To Identify Clustering Structure) is a clustering algorithm that
+/// doesn't explicitly cluster the data but instead creates an "augmented ordering" of the dataset
+/// representing it's density-based clustering structure. This ordering contains information which
+/// is equivalent to the density-based clusterings and can then be used for automatic and
+/// interactive cluster analysis.
+///
+/// OPTICS cluster analysis can be used to derive clusters equivalent to the output of other
+/// clustering algorithms such as DBSCAN. However, due to it's more complicated neighborhood
+/// queries it typically has a higher computational cost that other more specific algorithms.
+pub struct Optics;
+
+#[derive(Clone, Debug)]
+// This is a struct as in future may want to implement methods on it to get certain metrics from
+// the optics cluster distances.
+pub struct OpticsAnalysis<'a, F: Float> {
+    /// A list of the samples in the dataset sorted and with their reachability and core distances
+    /// computed.
+    pub orderings: Vec<Sample<'a, F>>,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Sample<'a, F: Float> {
+    /// Index of the sample in the dataset.
+    pub index: usize,
+    /// A reference to the datum.
+    pub observation: ArrayView<'a, F, Ix1>,
+    /// The reachability distance of a sample is the distance between the point and it's cluster
+    /// core or another point whichever is larger.
+    pub reachability_distance: Option<f64>,
+    /// The distance to the nth closest point where n is the minimum points to form a cluster.
+    pub core_distance: Option<f64>,
+}
+
+#[derive(Clone)]
+struct Neighbor<'a, F: Float> {
+    /// Index of the observation in the dataset
+    index: usize,
+    /// The observation
+    observation: ArrayView<'a, F, Ix1>,
+    /// The core distance
+    c_distance: Option<FloatOrd<f64>>,
```

Comment:
can you use standardized naming, i.e. `c_distance` or `core_distance`

Reply:
ah this is in conflict with the member right? I think `c_distance` should be o.k. with a comment above